### PR TITLE
MKL fixes

### PR DESCRIPTION
**Pull request by gf712**
I am not sure what causes this issue, but somehow separating evaluation of unary functions from the main expression fixes the issue.

I did some debugging and it seems like for example the square root operation of a matrix (`.cwiseSqrt()`) makes its rows and cols attributes be 0, which then causes an assertion error when this matrix is used in another operation.

This only happens with MKL backend and is first reported in #3992

### CODE REVIEW

**Comment by gf712** [comment]
If there are any Eigen experts around it would be nice to know what is going wrong here so we can file a ticket at Eigen. From what I can gather is that unary functions (power and sqrt) are not being properly evaluated lazily and have to be force evaluated for these code snippets to work.

File: src/shogun/converter/ica/SOBI.cpp [hunk 1]
```diff
@@ -66,8 +66,11 @@ void CSOBI::fit_dense(CDenseFeatures<float64_t>* features)
 	MatrixXd M0 = cor(EX,int(m_tau[0]));
 	EigenSolver<MatrixXd> eig;
 	eig.compute(M0);
-	MatrixXd SPH = (eig.pseudoEigenvectors() * eig.pseudoEigenvalueMatrix().cwiseSqrt() * eig.pseudoEigenvectors ().transpose()).inverse();
-	MatrixXd spx = SPH*EX;
+	MatrixXd EVMsqrt = eig.pseudoEigenvalueMatrix().cwiseSqrt();
```
Comment (karlnapf) [inline src/shogun/converter/ica/SOBI.cpp]: just asking, is this covered by any tests that would detect changing results?

Comment (gf712) [reply src/shogun/converter/ica/SOBI.cpp]: Should be covered [here](https://github.com/shogun-toolbox/shogun/blob/41888fe7c8dc3797063d674f452e13351f321338/tests/unit/converter/ica/SOBI_unittest.cc). That's how I found these bugs!

Comment (gf712) [reply src/shogun/converter/ica/SOBI.cpp]: Is there a buildbot for MKL backend btw?

**Comment by vigsterkr** [comment]
Add one to azure :)

**Comment by gf712** [comment]
> Add one to azure :)
> […](#)

OK! Just for GCC should be fine right? Might need some help setting it up.

**Comment by vigsterkr** [comment]
Yep but use clang as its faster compiling;)
I’ll b on irc in a bit

**Comment by vigsterkr** [comment]
closing in favour of #4529
@gf712 right?

**Comment by gf712** [comment]
Yup!

**Pull request closed without merging**